File: src/main/java/io/github/tanguygab/conditionalactions/conditions/ConditionManager.java
package io.github.tanguygab.conditionalactions.conditions;

import io.github.tanguygab.conditionalactions.ConditionalActions;
import io.github.tanguygab.conditionalactions.Utils;
import io.github.tanguygab.conditionalactions.conditions.types.*;
import org.bukkit.OfflinePlayer;
import org.bukkit.configuration.ConfigurationSection;

import java.util.*;
import java.util.function.Function;

public class ConditionManager {

    private final Map<String, ConditionGroup> conditions = new HashMap<>();
    private final Map<String, ConditionalOutput> conditionalOutputs = new HashMap<>();

    private final Map<String, Function<String,ConditionType>> types = new LinkedHashMap<>() {{
        put(">=", input -> new NumericCondition(input.split(">="), (left, right) -> left >= right));
        put(">", input -> new NumericCondition(input.split(">"), (left, right) -> left > right));
        put("<=", input -> new NumericCondition(input.split("<="), (left, right) -> left <= right));

        put("<-", input -> new StringCondition(input.split("<-"), String::contains));
        put("-|", input->new StringCondition(input.split("-\\|"), String::endsWith));
        put("|-", input->new StringCondition(input.split("\\|-"), String::startsWith));
        put("==", input->new StringCondition(input.split("=="), String::equals));
        put("=", input->new StringCondition(input.split("="), String::equalsIgnoreCase));

        put("permission:", input->new ConditionType(input.split("permission:")) {
            @Override
            public boolean isMet(OfflinePlayer p, Map<String, String> replacements) {
                String permission = ConditionalActions.parseReplacements(rightSide,replacements);
                return p != null && p.getPlayer() != null && p.getPlayer().hasPermission(permission);
            }
        });
    }};

    @SuppressWarnings("unchecked")
    public ConditionManager(ConditionalActions plugin) {
        Utils.updateFiles(plugin, "conditions.yml","conditions/default-conditions.yml");

        Utils.loadFiles("conditions", (key, value) -> {
            if (!(value instanceof ConfigurationSection section)) return;

            if (key.equalsIgnoreCase("conditions")) {
                section.getValues(false).forEach((name,cfg)->{
                    StringBuilder builder = new StringBuilder();
                    if (cfg instanceof List<?> list) {
                        list.forEach(obj -> {
                            if (obj instanceof List<?> andConditions)
                                builder.append(String.join("&&", (List<String>) andConditions));
                            else builder.append(obj);
                            builder.append("||");
                        });
                    } else builder.append(cfg);
                    if (builder.length() >= 2 && builder.substring(builder.length()-2).equals("||"))
                        builder.delete(builder.length()-2, builder.length());

                    conditions.put(name, new ConditionGroup(this, builder.toString()));
                });
                return;
            }
            if (key.equalsIgnoreCase("conditional-outputs")) {
                section.getValues(false).forEach((name,cfg)->{
                    if (cfg instanceof ConfigurationSection subsection)
                        conditionalOutputs.put(name, new ConditionalOutput(this, subsection));
                });
            }
        });
    }

    public ConditionGroup getCondition(String condition) {
        return conditions.computeIfAbsent(condition,k->new ConditionGroup(this,condition));
    }

    public ConditionalOutput getConditionalOutput(String name) {
        return conditionalOutputs.get(name);
    }

    public ConditionType find(String condition) {
        String condition0 = condition.startsWith("!") ? condition.substring(1) : condition;
        if (conditions.containsKey(condition0))
            return new GroupCondition(conditions.get(condition0),condition.startsWith("!") ? "!" : "");

        for (String separator : types.keySet())
            if (condition.contains(separator))
                return types.get(separator).apply(condition);
        return null;
    }

    public Set<String> getConditions() {
        return conditions.keySet();
    }
}

File: src/main/kotlin/io/github/tanguygab/conditionalactions/conditions/ConditionManager.kt
package io.github.tanguygab.conditionalactions.conditions

import io.github.tanguygab.conditionalactions.ConditionalActions
import io.github.tanguygab.conditionalactions.Utils
import io.github.tanguygab.conditionalactions.conditions.types.ConditionType
import io.github.tanguygab.conditionalactions.conditions.types.GroupCondition
import io.github.tanguygab.conditionalactions.conditions.types.NumericCondition
import io.github.tanguygab.conditionalactions.conditions.types.StringCondition
import me.neznamy.tab.api.event.plugin.TabLoadEvent
import me.neznamy.tab.shared.TAB
import org.bukkit.OfflinePlayer
import org.bukkit.configuration.ConfigurationSection
import org.bukkit.configuration.file.YamlConfiguration
import org.bukkit.entity.Player

class ConditionManager(plugin: ConditionalActions) {
    private val conditions = mutableMapOf<String, ConditionGroup>()
    private val conditionalOutputs = mutableMapOf<String, ConditionalOutput>()

    private val types = mutableMapOf<String, (String) -> ConditionType>().apply {
        put("<-") { StringCondition(it.split(" *<- *".toRegex(), limit = 2)) { left: String, right: String -> left.contains(right) } }
        put("-|") { StringCondition(it.split(" *-\\| *".toRegex(), limit = 2)) { left: String, right: String -> left.endsWith(right) } }
        put("|-") { StringCondition(it.split(" *\\|- *".toRegex(), limit = 2)) { left: String, right: String -> left.startsWith(right) } }

        put(">=") { NumericCondition(it.split(" *>= *".toRegex(), limit = 2)) { left: Double, right: Double -> left >= right } }
        put(">") { NumericCondition(it.split(" *> *".toRegex(), limit = 2)) { left: Double, right: Double -> left > right } }
        put("<=") { NumericCondition(it.split(" *<= *".toRegex(), limit = 2)) { left: Double, right: Double -> left <= right } }
        put("<") { NumericCondition(it.split(" *> *".toRegex(), limit = 2)) { left: Double, right: Double -> left < right } }

        put("!==") { StringCondition(it.split(" *!== *".toRegex(), limit = 2)) { left: String, right: String -> left != right } }
        put("!=") { StringCondition(it.split(" *!= *".toRegex(), limit = 2)) { left: String, right: String -> !left.equals(right, ignoreCase = true) } }

        put("==") { StringCondition(it.split(" *== *".toRegex(), limit = 2)) { left: String, right: String -> left == right } }
        put("=") { StringCondition(it.split(" *= *".toRegex(), limit = 2)) { left: String, right: String -> left.equals(right, ignoreCase = true) } }


        put("permission:") { object : ConditionType(it.split("permission: *".toRegex(), limit = 2)) {
                override fun isMet(player: OfflinePlayer?): Boolean {
                    val permission = rightSide
                    return player is Player && player.hasPermission(permission)
                }
            }
        }
    }

    init {
        Utils.updateFiles(plugin, "conditions.yml", "conditions/default-conditions.yml")

        Utils.loadFiles(plugin.dataFolder.resolve("conditions"), "") { file, _ ->
            YamlConfiguration.loadConfiguration(file).getValues(false).forEach { (key, value) ->
                if (value !is ConfigurationSection) return@forEach

                if (key.equals("conditions", ignoreCase = true)) {
                    value.getValues(false).forEach { (name: String, cfg: Any) ->
                        val builder = StringBuilder()
                        if (cfg is MutableList<*>) {
                            cfg.forEach { obj ->
                                @Suppress("UNCHECKED_CAST")
                                builder.append(if (obj is MutableList<*>) (obj as MutableList<String>).joinToString("&&") else obj)
                                builder.append("||")
                            }
                        } else builder.append(cfg)
                        if (builder.length >= 2 && builder.substring(builder.length - 2) == "||")
                            builder.delete(builder.length - 2, builder.length)
                        conditions[name] = ConditionGroup(this, builder.toString(), name)
                    }
                    return@forEach
                }
                if (key.equals("conditional-outputs", ignoreCase = true)) {
                    value.getValues(false).forEach { (name: String, cfg: Any) ->
                        if (cfg is ConfigurationSection) conditionalOutputs[name] = ConditionalOutput(this, cfg)
                    }
                }
            }
        }

        if (plugin.server.pluginManager.isPluginEnabled("TAB")) {
            TAB.getInstance().eventBus?.register(TabLoadEvent::class.java) {
                conditions.values.forEach { it.registerTABPlaceholder() }
            }
        }
    }

    fun getCondition(condition: String) = conditions.computeIfAbsent(condition) { ConditionGroup(this, condition) }

    fun getConditionalOutput(name: String) = conditionalOutputs[name]

    fun find(condition: String): ConditionType? {
        val inverted = condition.startsWith("!")
        val condition0 = condition.substring(if (inverted) 1 else 0)
        if (condition0 in conditions) return GroupCondition(conditions[condition0]!!, condition)

        for (separator in types.keys) if (condition.contains(separator)) return types[separator]!!(condition)
        return null
    }

    fun getConditions() = conditions.keys
}
